File: searchingAndSorting/peakElement/code.java
/*
gfg : https://practice.geeksforgeeks.org/problems/peak-element/1/?track=dsa-workshop-1-search-sort&batchId=308#


A peak element in an array is the one that is not smaller than its neighbours.
Given an array of size N, find the index of any one of its peak elements.
 

Example 1:

Input:
N = 3
arr[] = {1,2,3}
Output: 2 
Explanation: index 2 is 3.
It is the peak element as it is 
greater than its neighbour 2.
 

Example 2:

Input:
N = 2
arr[] = {3,4}
Output: 1
Explanation: 4 (at index 1) is the 
peak element as it is greater than 
its only neighbour element 3.
 

Your Task:
You don't have to read input or print anything. Complete the function peakElement() which takes the array arr[] and its size N as input parameters and return the index of any one of its peak elements.
Note: The generated output will be 1 if the index that you return is correct. Otherwise output will be 0. 

 

Expected Time Complexity: O(log N)
Expected Auxiliary Space: O(1)

 

Constraints:
1 <= N <= 10^5
1 <= A[] <= 10^6
*/

int peakElement(int arr[], int n)
{
   // Your code here
   if(n==1)
    return arr[0];
  if(arr[0]>arr[1])
        return 0;
    if(arr[n-1]>=arr[n-2])
        return n-1;
    int lo = 0;
    int hi = n-1;
    while(lo<=hi){
        int mid = (lo+hi)/2;
        if(arr[mid+1]<=arr[mid] && arr[mid]>=arr[mid-1])
            return mid;
        else if(arr[mid+1]>=arr[mid])
            lo = mid+1;
        else if(arr[mid-1]>=arr[mid])
            hi = mid-1;
    }
    return -1;
}

File: searchingAndSorting/binarySearch/code.cpp
/*
gfg : https://practice.geeksforgeeks.org/problems/binary-search-1587115620/1/?track=dsa-workshop-1-search-sort&batchId=308#

Given a sorted array of size N and an integer K, find the position at which K is present in the array using binary search.


Example 1:

Input:
N = 5
arr[] = {1 2 3 4 5} 
K = 4
Output: 3
Explanation: 4 appears at index 3.

Example 2:

Input:
N = 5
arr[] = {11 22 33 44 55} 
K = 445
Output: -1
Explanation: 445 is not present.

Your Task:  
You dont need to read input or print anything. Complete the function binarysearch() which takes arr[], N and K as input parameters and returns the index of K in the array. If K is not present in the array, return -1.


Expected Time Complexity: O(LogN)
Expected Auxiliary Space: O(LogN) if solving recursively and O(1) otherwise.


Constraints:
1 <= N <= 10^4
1 <= arr[i] <= 10^4
*/

int bin_search(int A[], int left, int right, int k)
{
    while(left<=right){
        int mid = (left+right)/2;
        if(k>A[mid])
            left = mid+1;
        else if(k<A[mid])
            right = mid-1;
        else
            return mid;
    }
    return -1;
}
